### PR TITLE
add client registration to the admin API

### PR DESCRIPTION
**Pull request by ericchiang**
work for #326

Need to add documentation and add generated schemas

### CODE REVIEW

**Comment by ericchiang** [comment]
@bobbyrullo ready for review

File: admin/api.go [hunk 1]
```diff
@@ -11,22 +16,32 @@ import (
...
+	userManager        *manager.UserManager
+	userRepo           user.UserRepo
+	passwordInfoRepo   user.PasswordInfoRepo
+	clientIdentityRepo client.ClientIdentityRepo
+	localConnectorID   string
 }
 
-func NewAdminAPI(userManager *manager.UserManager, userRepo user.UserRepo, pwiRepo user.PasswordInfoRepo, localConnectorID string) *AdminAPI {
+// TODO(ericchiang): Swap the DbMap for a storage interface. See #278
+
```
Comment (bobbyrullo) [inline admin/api.go]: I really don't like the DbMap leaking out of where it's supposed to be, so I am going to raise the priority of #385

File: admin/api.go [hunk 2]
```diff
@@ -11,22 +16,32 @@ import (
...
 	if localConnectorID == "" {
 		panic("must specify non-blank localConnectorID")
 	}
 
+	userRepo := db.NewUserRepo(dbMap)
+	passwordInfoRepo := db.NewPasswordInfoRepo(dbMap)
+	connectorConfigRepo := db.NewConnectorConfigRepo(dbMap)
+	txFactory := db.TransactionFactory(dbMap)
+	opts := manager.ManagerOptions{}
+	userManager := manager.NewUserManager(userRepo, passwordInfoRepo, connectorConfigRepo, txFactory, opts)
```
Comment (bobbyrullo) [inline admin/api.go]: why create the manager here? that seems like a step in the wrong direction; ideally, APIs would talk only to managers not repos.

File: admin/api.go [hunk 3]
```diff
@@ -108,6 +123,33 @@ func (a *AdminAPI) GetState() (adminschema.State, error) {
...
+	if err != nil {
+		return oidc.ClientRegistrationResponse{}, mapError(err)
+	}
+	c, err := a.clientIdentityRepo.New(id, req.Client)
+	if err != nil {
+		return oidc.ClientRegistrationResponse{}, mapError(err)
+	}
+	if req.IsAdmin {
+		// TODO(ericchiang): Put this in a transaction
+		if err := a.clientIdentityRepo.SetDexAdmin(c.ID, req.IsAdmin); err != nil {
```
Comment (bobbyrullo) [inline admin/api.go]: please do that in this PR

File: admin/api.go (hunk 2 above)
Comment (ericchiang) [inline admin/api.go]: Way fewer arguments to "NewAdminAPI". Makes it way easier to create the AdminAPI.

File: pkg/gendoc/markdown.go [hunk 4]
```diff
@@ -60,7 +60,7 @@ __Version:__ {{ .Version }}
```
Comment (bobbyrullo) [inline pkg/gendoc/markdown.go]: can you change the comment to accurately reflect the contents of the commit? "updated generated files: documentation, API etc."

File: admin/api.go (hunk 3 above)
Comment (ericchiang) [inline admin/api.go]: Will make `clientIdentityRepo.New` take a `is_admin` argument and do that there.

File: admin/api.go (hunk 2 above)
Comment (bobbyrullo) [inline admin/api.go]: ideally, APIs would just get managers and not talk to repos at all....but that's not where we are at yet. Still, no other API does this and I feel like it's a step in the wrong direction. Also, this makes it so you can't set the managers clock which is useful in tests.

File: integration/admin_api_test.go [hunk 5]
```diff
@@ -252,6 +259,73 @@ func TestCreateAdmin(t *testing.T) {
...
+			req := struct {
+				Client  oidc.ClientMetadata `json:"client"`
+				IsAdmin bool                `json:"isAdmin"`
+			}{tt.client, true}
+
+			data, err := json.Marshal(&req)
+			if err != nil {
+				return err
+			}
+			resp, err := f.hc.Post(s.URL+"/api/v1/client", "application/json", bytes.NewReader(data))
```
Comment (bobbyrullo) [inline integration/admin_api_test.go]: use the admin client to test - that way you know you've implemented the API correctly. See TestGetState for example.

File: pkg/gendoc/markdown.go (hunk 4 above)
Comment (bobbyrullo) [inline pkg/gendoc/markdown.go]: also please dont mix generated and handwritten code - the v1.json is in this commit and it should not be.

File: schema/adminschema/v1.json [hunk 6]
```diff
@@ -44,7 +44,26 @@
                   "type": "boolean"
               }
           }
-      }
+      },
+      "ClientCreateRequest": {
```
Comment (bobbyrullo) [inline schema/adminschema/v1.json]: ClientCreate? Or ClientRegister - choose one! :)

File: schema/adminschema/v1.json [hunk 7]
```diff
@@ -44,7 +44,26 @@
...
+          "id": "ClientCreateRequest",
+          "type": "object",
+	      "description": "'client' field is a client registration request as defined by the OpenID Connect dynamic registration spec, and holds fields such as redirect URLs, prefered algorithms, etc. For brevity field names and types of that object have been omitted.",
+          "properties": {
```
Comment (bobbyrullo) [inline schema/adminschema/v1.json]: the client object is necessary or else the generated client doesn't work right?

Comment (ericchiang) [inline schema/adminschema/v1.json]: Seems to work okay.

https://github.com/ericchiang/dex/blob/admin_client_registration/schema/adminschema/README.md#clientcreaterequest

Comment (bobbyrullo) [inline schema/adminschema/v1.json]: I don't mean the generated code. I mean the generated client library.

Comment (ericchiang) [inline schema/adminschema/v1.json]: It creates an empty struct. Any idea how to make it an `interface{}` or something similar?

File: admin/api.go (hunk 3 above)
Comment (bobbyrullo) [inline admin/api.go]: or just have SetDexAdmin take an tx

Comment (ericchiang) [inline admin/api.go]: Would have to update `New`, and `SetDexAdmin` to take one as well and add a transaction factory to AdminAPI. But either way.

File: schema/adminschema/v1.json (hunk 7 above)
Comment (bobbyrullo) [inline schema/adminschema/v1.json]: Nope, and that wouldn't be a great API either, taking an `interface{}` when always expects a certain struct.

 Maybe this API doesn't have to have the full OIDC dynamic reg stuff. Or maybe we reimplement it all in that JSON. 

The main thing is that the generated client code should work.

Comment (ericchiang) [inline schema/adminschema/v1.json]: Okay. We need to document that these exist btw; even I didn't know this generated a client. Would be a pity if we did all this work and no one used them.

**Comment by ericchiang** [comment]
@bobbyrullo I've got an update but it requires force pushing (had to undo the `NewAdminAPI` stuff). Want me to open an new PR or force push here?

**Comment by ericchiang** [comment]
Work is here https://github.com/ericchiang/dex/tree/admin_client_registration_2

**Comment by ericchiang** [comment]
Gentle bump

**Comment by ericchiang** [comment]
closing in favor of #392

**Pull request closed without merging**